tests/attention/test_hopper_fp8_attention.py

```diff
@@ -183,7 +183,248 @@ def test_block_sparse_attention(
+@pytest.mark.parametrize("batch_size", [2, 4])
+@pytest.mark.parametrize("num_heads", [8, 32])
+@pytest.mark.parametrize("head_dim", [64, 128, 256])
+@pytest.mark.parametrize("causal", [True, False])
+@pytest.mark.parametrize("dtype", [torch.float8_e4m3fn, torch.float8_e5m2])
+def test_batch_prefill_ragged(batch_size, num_heads, head_dim, causal, dtype):
+    if not is_sm90a_supported(torch.device("cuda")):
+        pytest.skip("SM90A is not supported")
+
+    print(
+        f"Testing FP8 batch prefill ragged with batch_size={batch_size}, num_heads={num_heads}, "
+        f"head_dim={head_dim}, causal={causal}, dtype={dtype}"
+    )
+
+    # Setup
+    o_dtype = torch.half
+    num_qo_heads = num_kv_heads = num_heads
+
+    # Create variable length sequences
+    torch.manual_seed(0)
+    qo_lens = [128 * (i + 1) for i in range(batch_size)]
+    kv_lens = [128 * (i + 1) for i in range(batch_size)]
+
+    # Build ragged tensors
+    qo_indptr = torch.tensor(
+        [0] + [sum(qo_lens[: i + 1]) for i in range(batch_size)],
+        dtype=torch.int32,
+        device="cuda",
+    )
+    kv_indptr = torch.tensor(
+        [0] + [sum(kv_lens[: i + 1]) for i in range(batch_size)],
+        dtype=torch.int32,
+        device="cuda",
+    )
+
+    total_qo_len = sum(qo_lens)
+    total_kv_len = sum(kv_lens)
+
+    # Create input tensors (fp16)
+    q_fp16 = torch.randn(
+        total_qo_len, num_qo_heads, head_dim, dtype=torch.half, device="cuda"
+    )
+    k_fp16 = torch.randn(
+        total_kv_len, num_kv_heads, head_dim, dtype=torch.half, device="cuda"
+    )
+    v_fp16 = torch.randn(
+        total_kv_len, num_kv_heads, head_dim, dtype=torch.half, device="cuda"
+    )
+
+    # Get reference output using fp16
+    wrapper_fp16 = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(
+        torch.empty(128 * 1024 * 1024, dtype=torch.uint8, device="cuda"),
+        "NHD",
+        backend="fa3",
+    )
+    wrapper_fp16.plan(
+        qo_indptr,
+        kv_indptr,
+        num_qo_heads,
+        num_kv_heads,
+        head_dim,
+        head_dim,
+        causal=causal,
+    )
+    o_ref = wrapper_fp16.run(q_fp16, k_fp16, v_fp16)
+
+    # Quantize to FP8
+    q_fp8, s_q = per_head_symmetric_quant(q_fp16, quant_dtype=dtype)
+    k_fp8, s_k = per_head_symmetric_quant(k_fp16, quant_dtype=dtype)
+    v_fp8, s_v = per_head_symmetric_quant(v_fp16, quant_dtype=dtype)
+
+    # Run FP8 batch prefill with ragged KV cache
+    wrapper_fp8 = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(
+        torch.empty(128 * 1024 * 1024, dtype=torch.uint8, device="cuda"),
+        "NHD",
+        backend="fa3",
+    )
+    wrapper_fp8.plan(
+        qo_indptr,
+        kv_indptr,
+        num_qo_heads,
+        num_kv_heads,
+        head_dim,
+        head_dim,
+        q_data_type=dtype,
+        kv_data_type=dtype,
+        o_data_type=o_dtype,
+        causal=causal,
+    )
+    o_fp8 = wrapper_fp8.run(q_fp8, k_fp8, v_fp8, s_q, s_k, s_v)
```

Comment:
We should have tests that cover both types of scaling factors:

1. host-side per-tensor scales
2. device-side per-head scale tensors